### ISSUE TITLE
Provide IntelliSense for PortalScripting

### ISSUE THREAD

**Issue by tineheller** [otris/vscode-janus-debug#45]
The DOCUMENTS application server supports an embedded server side scripting language called PortalScript. The PortalScript runtime environment grants access to a couple of properties of files or fields. Furthermore, the DOCUMENTS scripting interface (PortalScripting API) supports several functions and predefined objects (e.g. context and util). These predefined functions and objects are implicitly available in every script that is executed on DOCUMENTS server.

**Get IntelliSense for the entire PortalScripting API manually via ```.d.ts``` files**
As a first step, we will simply provide suitable ```.d.ts``` files. So you can get IntelliSense by downloading these files into your project and adding the path to these files to your ```jsconfig.json``` or only adding an empty ```jsconfig.json``` to your project.

**Get IntelliSense automatically**
But one of the nice features in VS Code is getting IntelliSense without caring about it. So we would like to provide IntelliSense for PortalScripting automatically.
For all I know there are two ways in VS Code to provide IntelliSense:

1. Creating a Language Server. But this might be overkill.
2. Loading the ```d.ts``` files and writing into the ```jsconfig.json``` automatically with the extension. But this seems rather to be a hack because then the extension will add and change files in the users' projects again. We really want to reduce that.

**Comment by tineheller** [comment]
I think this question on Stackoverflow describes exactly what I want to do.

https://stackoverflow.com/questions/39698336/inject-implicit-ambient-typings-via-visual-studio-code-extension

The answer is, to write the d.ts file to the user's project and add an entry to the jsconfig.json or tsconfig.json. Even though it might look like a hack it seems to be the way how an extension can extend intellisense for JavaScript.

**Comment by tineheller** [comment]
As a compromise I could add a command like ```"Get PortalScript Intellisense"``` to the extension. And then ask the user where to copy the ```.d.ts``` files and ask for permission to write in the user's ```jsconfig.json```.

**Comment by FabianTe** [comment]
@tineheller Make it ``Add PortalScript Intellisense`` and you're good to go

**Comment by tineheller** [comment]
@bkircher What do you think?

**Comment by bkircher** [comment]
Well, I just got a couple of questions.

* What do you do if there is already a `jsconfig.json` file present in the project folder?
* Where do you install the typings to?
* How do you go about different versions of the typings (for example when the API changes between DOCUMENTS versions)?
* How do you differentiate between JANUS (which has it's own "ambient" definitions of functions and objects) and DOCUMENTS?
* What will the user do with those files in his or her VCS? Ignore them, add them?
* Will those typings ever get an update? If yes, how?

**Comment by tineheller** [comment]
- If there is already a `jsconfig.json`, I will add an entry to the `include` property. I will ask the user for permission to do that.
- As described above, I will ask the user where to copy the typings and suggest the project root.
- I could also ask the user about the DOCUMENTS version and suggest the latest version.
- I could install the typings with the extension and copy them from there. So an updated version of the typings will come with an extension update.
- Maybe I should create a new issue for the problem about how to differentiate between Janus and DOCUMENTS. For now there will be no differentiation.
- I think the users should ignore the files in their VCS.

**Comment by FabianTe** [comment]
> As described above, I will ask the user where to copy the typings and suggest the project root

As with the saved password (#49): I'd rather save such files in a plugin folder to keep the project folder clean. Ignoring one plugin folder in VCSs is always more user friendly than having to ignore multiple (possibly hidden) plugin/extension files in the project root folder.

> I could install the typings with the extension and copy them from there. So an updated version of the typings will come with an extension update.

If done like this you could save all typings versions locally and have the user choose which version he wants to use. Would be neat!

**Comment by tineheller** [comment]
@bkircher We don't need to differentiate between DOCUMENTS and Janus. Because only intellisense for DOCUMENTS PortalScripting will be supported.

**Comment by bkircher** [comment]
> If there is already a jsconfig.json, I will add an entry to the include property. I will ask the user for permission to do that.

snip

>> Where do you install the typings to?

> As described above, I will ask the user where to copy the typings and suggest the project root.


Thanks for the answers, @tineheller 

tl;dr: I don't think this is a good idea. 

Long version: I think what we really want is to have it work automatically, without the user having to put files into his project. We want a VS Code extension that allows the user to have IntelliSense for those "ambient" objects and functions out-of-the-box, as soon the extension is activated. No copying around of files and asking the user for permission to change configuration files in his project, making assumptions about his directory layout and such.

Until we have something providing a _out-of-the-box experience_ (if we will ever): The user should copy those typings into his project manually by himself. We provide a Wiki page where the steps to get IntelliSense working are described.

1. Download this: https://raw.githubusercontent.com/otris/vscode-janus-debug/master/portalscript/typings/portalScripting.d.ts into your project
2. Create file `jsconfig.json` with following contents: `{ /* ... */ }`
3. And voilà: it works


> We don't need to differentiate between DOCUMENTS and Janus. Because only intellisense for DOCUMENTS PortalScripting will be supported.

This would leave some of us w/o IntelliSense for JANUS at all. Why would we want something like that?

**Comment by bkircher** [comment]
Is it possible to inject typings into the typescript language server via extension? Maybe we could skim through: this https://github.com/Microsoft/TypeScript/tree/master/src/server ?

**Comment by FabianTe** [comment]
@bkircher 
Wrote a [wiki ](https://github.com/FabianTe/vscode-janus-debug/wiki/Add-DOCUMENTS-intellisense)article. Is there a way to push that from my fork to yours?

**Comment by bkircher** [comment]
> Wrote a wiki article. Is there a way to push that from my fork to yours?

Wow, great! Do you mind adding the page directly to our project Wiki?

I added you to the collaborators list. (You should've received an email from GitHub. This way you are allowed to make changes in the Wiki.)

I suppose the usual Ctrl+C, Ctrl-V workflow is quickest here :wink:

**Comment by FabianTe** [comment]
Thanks. I removed my comment because my wiki contained a fault, I'll add the corrected version as soon as possible.

**Comment by m-Schlitzer** [comment]
With the "typings" folder I got from @tineheller and a jsconfig.json containing 
```
{
    "include": [
        "bower_components/**/*",
        "src/**/*",
	"typings/**/*"
    ]
}
```

I got IntelliSense working for file specific and global imports.

**Comment by tineheller** [comment]
@m-Schlitzer What do you mean? You got IntelliSense working for imports?

**Comment by m-Schlitzer** [comment]
Well, yes. IntelliSense seems to be working correctly now with this configuration, letting me see the correct imported functions.

**Comment by FabianTe** [comment]
For me it is working right now too! :+1: 

Wiki article: https://github.com/otris/vscode-janus-debug/wiki/IntelliSense-for-PortalScripting

![image](https://user-images.githubusercontent.com/7669818/27575886-75384f6c-5b1b-11e7-8902-4ef152d7b88d.png)

**Comment by bkircher** [comment]
Thanks for the article  @FabianTe

**Comment by idolum** [comment]
Works for me as well and for the time being I can live with it. But the downloading the d.ts file manually is something I don't like. If I got this right, you have to do this for every project. What about a command like "Get typings" donwloading the typings to the project? Just to make it more convenient for the users.

But as I said, for the time being it is ok and the issue has less priority from my point of view.

**Comment by bkircher** [comment]
I would prefer a "just-works" experience, too. A solution without the need to download files or otherwise manually add type definitions to my projects.

Also, I'd think that those type definitions may vary across projects and it would be nice if those definitions would match the DOCUMENTS versions and work for non-DOCUMENTS projects, too. This way, IntelliSense would not just be only "convenient" but could point out actual bugs in the code.

> But as I said, for the time being it is ok and the issue has less priority from my point of view.

I agree with that.

FWIW, an additional approach was pointed out to me on [gitter](https://gitter.im/Microsoft/vscode). We could use [Automatic Type Acquisition](https://code.visualstudio.com/docs/languages/javascript#_automatic-type-acquisition) (ATA). ATA is essentially what we want _except_ the `require()` part because our objects and classes are implicitly defined, so no need to `require()` them. But at least it would circumvent the manual "download" step. Only thing we need to do for this is to provide the type definitions as a npm package.

I guess this pretty much leaves us with:

a) Force users to have one of those project files, `jsconfig.json`, `tsconfig.json`, `package.json` and do download or otherwise add the type definitions to each of their projects.

b) Open an issue with VS Code explaining our use case and asking them to provide a VS Code configuration setting that would do the same as "automatic type acquisition", but work in all JS files even without a `require()` -- bonus points if we offer to implement it ourselves.

c) Open an issue with VS Code asking them to provide a hook for the [language server](https://github.com/Microsoft/TypeScript/tree/master/src/server) that would allow types to be added by an extension. Again, bonus points when we support that with a PR.

d) Dig into the [code](https://github.com/Microsoft/TypeScript/tree/master/src/server) and see if there is something I have missed :grinning:

If I have missed something, though, hopefully somebody will pop around and enlighten us.

**Comment by FabianTe** [comment]
> FWIW, an additional approach was pointed out to me on gitter. We could use Automatic Type Acquisition (ATA). ATA is essentially what we want except the require() part because our objects and classes are implicitly defined, so no need to require() them.

This might come in handy here: **Search #import'ed scripts to improve IntelliSense #32**

**Comment by tineheller** [comment]
> c) Open an issue with VS Code asking them to provide a hook for the language server that would allow types to be added by an extension. Again, bonus points when we support that with a PR.

I'd like that.

**Comment by FabianTe** [comment]
What about adding publishing a definition on [NPM Types](https://www.npmjs.com/~types)?

**Comment by bkircher** [comment]
> What about adding publishing a definition on NPM Types?

Yes, we considered that, too, if I remember correctly.

This would allow for VS Code's Automatic Type Acquisition and thus spare us from downloading typings ourselves. Plus, typings would be versioned, making it transparent for the user and allow for possibly different typings depending on the DOCUMENTS server version. Updating typings would also be possible and transparent for the user.

That being said, I don't follow this discussion much anymore. The decision-making process got a bit confusing with too many people involved (not in this thread) and no clear path to move forward.

Sadly, because IMO this is one of the most interesting and important features of this extension.

**Comment by FabianTe** [comment]
> Sadly, because IMO this is one of the most interesting and important features of this extension.

I agree.

> That being said, I don't follow this discussion much anymore. The decision-making process got a bit confusing with too many people involved (not in this thread) and no clear path to move forward.

I guess you guys at otris would have to decide if you want to improve the typings situation or if you think it is enough as it is right now. If you wanted to improve it, consider adding it to npm types (because of the reasons you commented).

*Also consider locking this thread for now until you planned your next steps in regards with the topic*

**Comment by tineheller** [comment]
I'd like to close this issue because I think the ways that we discussed here are not really suitable to our cases. We have some special cases how and when users want to get IntelliSense.

**Example: file types (very dynamic)**

In addition to IntelliSense for PortalScripting, e.g. IntelliSense for file types is very helpful. But the file types depends on the DOCUMENTS server and they can be changed by the user at any time. For now only the user knows when they changed the file types. So it's important to be able to generate the typings via a command.

**Example: Gadgets, Script Extensions, DocumentsContext (only for advanced users)**

Additional there are some APIs like the Gadgets, the Script Extensions or the DocumentsContext. For this APIs the user has to now how to use them inside the portal script. And probably the user wants to decide which parts of all the available modules they want to use and which not. And that also can change at any time and the APIs depends on the DOCUMENTS server.

**Get IntelliSense from DOCUMENTS server**

So it seems it's best to get the suitable IntelliSense typings from the DOCUMENTS server and let the user decide which typings they want to use. This typings could also be provided automatically if the DOCUMENTS server is running, but additional it's important to let the user decide when to get the desired typings.

If required, a new issue can be opened.
